Add http(s)_proxy support in yum module

sillent opened the pull request:
##### SUMMARY
Fixes #18979

Added set_env_proxy function, that set system http(s)_proxy
environment for fetching RPM if proxy settings set in yum.conf file


##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
packaging/os/yum.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.6.0
```


##### ADDITIONAL INFORMATION
```

```

sillent committed:
added set_env_proxy function for setting proxy environment

Added set_env_proxy function, that set system http(s)_proxy
environment for fetching RPM if proxy settings set in yum.conf file
that fix Issue - #18979

sillent committed:
fix automatic field numbering specification

sillent committed:
changed if statement in setting http_proxy environment

mkrizek commented on lib/ansible/modules/packaging/os/yum.py:
```diff
@@ -639,6 +639,28 @@ def local_envra(path):
+        os.environ["{0}_proxy".format(schem)] = re.sub(r"({0}://)".format(schem),
+                                                       r"\1" + namepass, yumb.conf.proxy)
+    except AttributeError as ae:
+        raise ae
```
What is the point of catching the exception just to raise it again? Something is missing here probably?

sillent replied:
Sure... You right!
Maybe try/catch in that block of code doesn't matter of it

mkrizek commented:
Does this work just with the global yum's proxy from `yum.conf` or also with proxies defined in the repo files? I find it weird that this, IIUIC, changes system's `ENV` without changing it back to the initial state.

mkrizek commented on lib/ansible/modules/packaging/os/yum.py:
```diff
@@ -639,6 +639,27 @@ def local_envra(path):
+            os.environ["{0}_proxy".format(schem)] = re.sub(r"({0}://)".format(schem),
+                                                           r"\1" + namepass, yumb.conf.proxy)
+    except AttributeError as ae:
+        raise ae
```
Does this work just with the global yum's proxy from `yum.conf` or also with proxies defined in the repo files?

mkrizek replied:
What is the point of catching the exception just to raise it again? Something is missing here probably?

sillent committed:
Change set_env_proxy function to function with decorator

That decorator set system proxy environment from yum.conf and revert
back to system configuration after fetching RPM

sillent commented:
Ready for review

maxamillion commented on lib/ansible/modules/packaging/os/yum.py:
```diff
@@ -639,6 +639,42 @@ def local_envra(path):
 
 
+def set_env_proxy(function):
+    def fetching(spec, module, conf_file, installroot):
```
Out of curiosity, what's the motivation for making this a new (nested) function instead of adding the handling of the `yum.conf` proxy config items to `yum_base`?

sillent replied:
it does not matter, because in the methods of installation and update of the package, if the package name is specified as URI, then the external method - "fetch_url" is used, which works with the system settings of the proxy. And if system proxy is not set, or it's different from setted in yum.conf, than we need set back configuration to previous state. 

Because of that i wrote decorator with arguments, that temporary set or change system proxy environment, if in yum.conf set proxy variables, then package fetched with method  <code>def fetch_rpm_from_url(spec, module=None):</code>, and finaly we set previous state of system proxy environment

maxamillion replied:
Solid point about the specified URI, makes sense. Thanks. 👍

abadger replied:
This seems like it should be a context manager rather than a pseudo-decorator

abadger replied:
Note that it might also be easier to just include this logic with the existing fetch_rpm_from_url() function rather than using a context-manager.  I could see doing it either way.

abadger commented on lib/ansible/modules/packaging/os/yum.py:
```diff
@@ -639,6 +639,42 @@ def local_envra(path):
+        # setting system proxy environment and saving old, if exists
+        yumb = yum_base(conf_file, installroot)
+        namepass = ""
+        schem = "http"
```
s/schem/scheme/

sillent replied:
fix

abadger commented on lib/ansible/modules/packaging/os/yum.py:
```diff
@@ -639,6 +639,42 @@ def local_envra(path):
+        old_proxy_env = [os.getenv("http_proxy"), os.getenv("https_proxy")]
+        try:
+            if yumb.conf.proxy:
+                if 'https' in yumb.conf.proxy:
```
Should this be startswith instead of in?

sillent replied:
you right
fix

abadger commented on lib/ansible/modules/packaging/os/yum.py:
```diff
@@ -639,6 +639,42 @@ def local_envra(path):
+def set_env_proxy(function):
+    def fetching(spec, module, conf_file, installroot):
+        # setting system proxy environment and saving old, if exists
+        yumb = yum_base(conf_file, installroot)
```
I'm not sure why but other parts of the code are using "my" for the variable name for a yum_base object so we should keep doing that.

sillent replied:
fix

abadger commented on lib/ansible/modules/packaging/os/yum.py:
```diff
@@ -639,6 +639,42 @@ def local_envra(path):
+        except yum.Errors.YumBaseError as e:
+            module.fail_json(msg="Error setting/accessing proxy: %s" % to_native(e))
+        # getting package with new proxy environment
+        package = fetch_rpm_from_url(spec, module)
```
If you were to do this as a decorator, you should be calling ```function``` here instead of hardcoding fetch_rpm_from_url.

sillent replied:
You mean using 
```
@set_env_prox(arg...)  
def fetch_rpm_from_urls(spec, module=None):
...
```
In that variant i cannot pass variables needed to YumBase 
> conf_file
> installroot

abadger replied:
I mean like:
``` python
-        package = fetch_rpm_from_url(spec, module)
+        package = function(spec, module)
```

But looking at this, I still think it should be a context manager rather than a pseudo-decorator.  That would look something like this:

``` python
from contextlib import contextmanager
[...]
@contextmanager
def set_env_proxy(conf_file, installroot):
    old_environ = # [old values of the proxy env vars]
    try:
        my = yum_base(conf_file, installroot)
        # [set environment to the proxies that yum knows about]
        yield
    except:
        raise
    finally:
        # reset the environment to the values saved in old_environ
[...]
if '://' in spec:
    with set_env_proxy(conf_file, installroot):
        package = fetch_rpm_from_url(spec, module=module)
[...]
```

sillent committed:
Minor fix

- rename variable schem  to scheme
- change 'in' to 'startswith'

sillent committed:
change decorator set_env_proxy to decorating through contextmanager

- added import contextmanager from contextlib
- set_env_proxy now decorating through contextmanager
- fix http/https setting environment principle

abadger commented:
@maxamillion, this is looking good to me now.  If it passes in shippable +1 from me to merge.
